implemented LDAP group pruning and added unit testing for LDAP group parts

stevekuznetsov opened the pull request:
Adds `openshift ex prune-groups`.

cc/ @deads2k @liggitt

danmcp commented:
@stevekuznetsov Needs rebase

stevekuznetsov commented:
@liggitt Tests added, ready for review
[test][extended:ldap_groups]

stevekuznetsov commented on pkg/cmd/experimental/syncgroups/ad/ldapinterface.go:
```diff
@@ -73,15 +73,17 @@ func (e *ADLDAPInterface) ExtractMembers(ldapGroupUID string) ([]*ldap.Entry, er
 	}
 
+	if len(usersInGroup) == 0 {
+		return nil, ldaputil.NewEntryNotFoundError()
```
I'm not completely happy exposing this error outside of the `ldaputil` package but unless we want to create our own version of it internally, we need this here.

stevekuznetsov commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,276 @@
+
+type PruneGroupsOptions struct {
+	// Config is the LDAP sync config read from file
+	Config api.LDAPSyncConfig
```
Chosen for compatibility, the entire sync config is not necessary for the prune job.

stevekuznetsov commented:
@deads2k  This is ready for review.

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -10,13 +10,41 @@ import (
+// Error returns the error string for the entry not found query
 func (e *errEntryNotFound) Error() string {
-	return "search for entry did not return any results"
+	if len(e.baseDN) > 0 {
```
No need for the check, just always output it.

deads2k commented on pkg/cmd/experimental/syncgroups/ad/ldapinterface.go:
```diff
@@ -73,15 +73,17 @@ func (e *ADLDAPInterface) ExtractMembers(ldapGroupUID string) ([]*ldap.Entry, er
 		}
 
-		for i := range currEntries {
-			currEntry := currEntries[i]
```
No.  This will break us.  The code here makes sure that you add a different pointer to the list for every entry.  The change you just made will assign them all to the same pointer, because `currEntry` in your code is a single spot in memory.

stevekuznetsov replied:
Doesn't seem to break anything. I also don't understand why these two constructs would have different output - one is syntactic sugar for the other.

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,276 @@
+	PruneGroupsRecommendedName = "prune-groups"
+
+	pruneGroupsLong = `
+Prune OpenShift Groups referencing records from an external provider.
```
referencing **deleted** records

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,276 @@
+been synced previously. Any combination of a literal whitelist, a whitelist file and a blacklist file is supported.
+The path to a sync configuration file that was used for syncing the groups in question is required in order to 
+describe how data is requested from the external record store. Default behavior is to prune all OpenShift groups 
+for which the LDAP record does not exist.
```
for which the external provider record does not exist

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,276 @@
+
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s [WHITELIST] [--whitelist=WHITELIST-FILE] [--blacklist=BLACKLIST-FILE] --sync-config=CONFIG-SOURCE", name),
+		Short:   "Prune OpenShift groups referencing records from an external provider.",
```
referencing deleted records

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,276 @@
+
+func (o *PruneGroupsOptions) Complete(whitelistFile, blacklistFile, configFile string, args []string, f *clientcmd.Factory) error {
+	// if args are given, they are OpenShift Group names forming a whitelist
+	if len(args) > 0 {
```
make a helper function for building the whitelist both here and for sync.

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,276 @@
+		o.Blacklist = append(o.Blacklist, blacklistData...)
+	}
+
+	yamlConfig, err := ioutil.ReadFile(configFile)
```
factor into a helper function.

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,276 @@
+		return err
+	}
+
+	if f != nil {
```
Remind me why this can be nil

stevekuznetsov replied:
This is inherited code from `oc secrets new` so I am not sure why, I think you wrote that code? We can remove the check if it can't be `nil`.

deads2k replied:
> This is inherited code from oc secrets new so I am not sure why, I think you wrote that code? We can remove the check if it can't be nil.

Remove it and see if you still work.

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,276 @@
+
+}
+
+func (o *PruneGroupsOptions) GetOpenShiftGroupListerMapper(syncBuilder SyncBuilder, clientConfig *ldaputil.LDAPClientConfig) (interfaces.LDAPGroupListerNameMapper, error) {
```
Fix these like we discussed in person.

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,91 @@
+	// Lists all groups to be synced
+	GroupLister interfaces.LDAPGroupLister
+	// Fetches a group and extracts object metainformation and membership list from a group
+	GroupMemberExtractor interfaces.LDAPMemberExtractor
```
Create a new interface who's job it is to detect whether a group exists or not and implement that with the ldap interfaces types

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,91 @@
+
+		// get membership data
+		_, err := s.GroupMemberExtractor.ExtractMembers(ldapGroupUID)
+		if ldaputil.IsNoSuchObjectError(err) || ldaputil.IsEntryNotFoundError(err) {
```
This detection is too fragile to satisfy me.  You'll need a new interface that does exactly what we want.

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,91 @@
+					continue
+				}
+			} else {
+				group, err := s.GroupClient.Get(groupName)
```
No.  Just output the names, don't bother re-getting it.

stevekuznetsov replied:
We don't want to follow the same paradigm of returning the list of groups so that on a dry run the command can work with that?

deads2k replied:
> We don't want to follow the same paradigm of returning the list of groups so that on a dry run the command can work with that?

We want to say what's being removed, but since we aren't mutating the object (other than deleting), there's no value in displaying it.

deads2k commented on test/extended/ldap_groups.sh:
```diff
@@ -226,5 +224,12 @@ for (( i=0; i<${#schema[@]}; i++ )); do
+	# PRUNING
+	echo -e "\tTEST: Sync all LDAP groups from LDAP server, change LDAP entries, then prune OpenShift groups"
+	openshift ex sync-groups --sync-config=sync-config.yaml --confirm || true
+	ldapmodify -x -h $LDAP_SERVICE_IP -p 389 -D cn=Manager,dc=example,dc=com -w admin -f prune.ldif
```
patch the `ldapGroupUID` of an OpenShift group.

stevekuznetsov commented:
@deads2k Addressed comments. Ready for review

stevekuznetsov commented:
[test][extended:ldap_groups]

deads2k commented on pkg/cmd/experimental/syncgroups/cli/augmented_ad.go:
```diff
@@ -8,20 +8,20 @@ import (
 )
 
-var _ SyncBuilder = &AugmentedADSyncBuilder{}
+var _ SyncBuilder = &AugmentedADBuilder{}
```
Also a `PruneBuilder`

deads2k commented on pkg/cmd/experimental/syncgroups/cli/interfaces.go:
```diff
@@ -0,0 +1,38 @@
+	GetBlacklist() []string
+}
+
+// OpenShiftGroupNameRestrictions describes an object that holds a Group client for OpenShift
```
bad comment

deads2k commented on pkg/cmd/experimental/syncgroups/cli/syncgroups.go:
```diff
@@ -174,23 +167,48 @@ func (o *SyncGroupsOptions) Complete(typeArg, whitelistFile, blacklistFile, conf
+}
+
+// buildNameLists builds the white- and blacklists
+func buildNameLists(source GroupSyncSource, args []string, whitelistFile, blacklistFile string) ([]string, []string, error) {
```
No.  `buildList(items []string, fileOfItems string) ([]string,  error)`

stevekuznetsov replied:
Ah, of course.

deads2k commented on pkg/cmd/experimental/syncgroups/ad/augmented_ldapinterface.go:
```diff
@@ -6,7 +6,7 @@ import (
 
 	"github.com/openshift/origin/pkg/auth/ldaputil"
-	ldapinterfaces "github.com/openshift/origin/pkg/cmd/experimental/syncgroups/interfaces"
+	"github.com/openshift/origin/pkg/cmd/experimental/syncgroups/interfaces"
```
`ldapinterfaces` so the package is consistent across all imports.  It makes reviewing code easier.

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -239,6 +260,9 @@ func QueryForEntries(clientConfig *LDAPClientConfig, query *ldap.SearchRequest)
 	glog.V(4).Infof("searching LDAP server %v://%v at dn=%q with scope %v for %s requesting %v", clientConfig.Scheme, clientConfig.Host, query.BaseDN, query.Scope, query.Filter, query.Attributes)
 	searchResult, err := connection.Search(query)
 	if err != nil {
+		if strings.Contains(err.Error(), "LDAP Result Code 32") {
```
The error we get back from upstream isn't typed?  Also, does this happen?  We don't just get back an empty result set (which is what I would expect.)

deads2k replied:
If this does happen, explain what conditions can cause it in comments here.

stevekuznetsov replied:
Upstream is not typed. This occurs when the `baseDN` of a search refers to an object that doesn't exist in the tree.

deads2k replied:
> Upstream is not typed. This occurs when the baseDN of a search refers to an object that doesn't exist in the tree.

Link to your upstream issue in this pull.

liggitt replied:
are you sure? Is it not this?:
```
type Error struct {
	Err        error
	ResultCode uint8
}
```

stevekuznetsov replied:
They're not individually typed, and I'm not about to do a switch-case on `uint8` values of their `Error` field. I've made an upstream pull (https://github.com/go-ldap/ldap/pull/36) to make identification easier.

deads2k commented on pkg/cmd/experimental/syncgroups/interfaces/interfaces.go:
```diff
@@ -35,3 +35,7 @@ type LDAPGroupListerNameMapper interface {
 	LDAPGroupNameMapper
 }
+
+type LDAPGroupLocator interface {
```
godoc

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,84 @@
+	for _, ldapGroupUID := range ldapGroupUIDs {
+		glog.V(1).Infof("Checking LDAP group %v", ldapGroupUID)
+
+		// get membership data
```
bad comment.  Comment probably isn't necessary.

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,84 @@
+		// get membership data
+		exists, err := s.GroupLocator.Exists(ldapGroupUID)
+		if err != nil {
+			fmt.Fprintf(s.Err, "Error determining LDAP group membership for %q: %v.\n", ldapGroupUID, err)
```
Bad output

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,84 @@
+			errors = append(errors, err)
+			continue
+		}
+		if !exists {
```
reverse this.  `if exists{ glog.V(1`); continue}`

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,84 @@
+			// if the LDAP entry that was previously used to create the group doesn't exist, prune it
+			groupName, err := s.GroupNameMapper.GroupNameFor(ldapGroupUID)
+			if err != nil {
+				errors = append(errors, err)
```
print to Err

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,84 @@
+			}
+
+			if !s.DryRun {
+				fmt.Fprintf(s.Out, "group/%s\n", groupName)
```
This output is unconditional.

stevekuznetsov replied:
It's conditional on `sync-groups`... we have the opposite of this inside of the `Run()` method for the `Options` structs to print when there is a dry run.

deads2k replied:
> It's conditional on sync-groups... we have the opposite of this inside of the Run() method for the Options structs to print when there is a dry run.

That relies on a printer that is damn near impossible to build manually, so we output in one big clump.  This command has no such need.

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,84 @@
+					continue
+				}
+			} else {
+				groupsToPrune = append(groupsToPrune, groupName)
```
This append should be unconditional.

deads2k commented on pkg/cmd/experimental/syncgroups/interfaces/errors.go:
```diff
@@ -1,8 +1,6 @@
-import (
-	"fmt"
-)
+import "fmt"
```
put it back in parens

stevekuznetsov replied:
Automated tools ...

deads2k replied:
back in parens

deads2k commented on test/extended/ldap_groups.sh:
```diff
@@ -225,5 +223,12 @@ for (( i=0; i<${#schema[@]}; i++ )); do
+	# PRUNING
+	echo -e "\tTEST: Sync all LDAP groups from LDAP server, change LDAP UID, then prune OpenShift groups"
+	openshift ex sync-groups --sync-config=sync-config.yaml --confirm
+	oc patch group ${group2_osuid} -p '{"metadata":{"annotations":{"openshift.io/ldap.uid":""}}}'
```
Set this to a value.

stevekuznetsov replied:
Why is this important? If anything, having no value is better because we will for sure be tripping the out-of-bounds error check.

deads2k replied:
> Why is this important? If anything, having no value is better because we will for sure be tripping the out-of-bounds error check.

Ok, then check both.

deads2k replied:
> Why is this important? If anything, having no value is better because we will for sure be tripping the out-of-bounds error check.

Because no admin is going to intentionally clear it and we're never going to create on that doesn't have it.  This is currently testing something that will never actually happen.  Not quite useless (as you pointed out), but darn close.

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+// Prune allows the LDAPGroupSyncer to be a GroupSyncer
+func (s *LDAPGroupPruner) Prune() ([]string, []error) {
```
Unit tests for this flow using mocks.

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,237 @@
+	// cmd.Flags().StringSliceVar(&options.Blacklist, "blacklist-group", options.Blacklist, "group to blacklist")
+	cmd.Flags().StringVar(&configFile, "sync-config", configFile, "path to the sync config")
+	cmd.Flags().BoolVar(&options.Confirm, "confirm", false, "if true, modify OpenShift groups; if false, display groups")
+	cmdutil.AddPrinterFlags(cmd)
```
These don't do anything, do they?

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,237 @@
+		return err
+	}
+	o.Whitelist = whitelist
+	blacklist, err := buildNameList(GroupSyncSourceOpenShift, []string{}, blacklistFile)
```
`o.Blacklist, err = buildnameList`

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,237 @@
+	return nil
+}
+
+func (o *PruneGroupsOptions) Validate() error {
```
We need a `GroupInterface` too

stevekuznetsov replied:
Why do we need to validate that? *Especially* now that the addition of the interface is not conditional on anything? If we fail to do it, `Complete` will error.

deads2k replied:
> Why do we need to validate that? Especially now that the addition of the interface is not conditional on anything? If we fail to do it, Complete will error.

Delegation.  If someone builds one of these, they might mess it up.  `Validate` will smack them.

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,87 @@
+		}
+
+		if s.DryRun {
+			groupsToPrune = append(groupsToPrune, groupName)
```
This shoudl be unconditional

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,87 @@
+		if s.DryRun {
+			groupsToPrune = append(groupsToPrune, groupName)
+		} else {
+			fmt.Fprintf(s.Out, "group/%s\n", groupName)
```
This shoudl be unconditional

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,233 @@
+
+	// Now we run the pruner and report any errors
+	groupsToPrune, pruneErrors := pruner.Prune()
+	if !o.Confirm {
```
This isn't needed.

stevekuznetsov commented:
@deads2k Addressed comments, added unit tests

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -10,13 +10,34 @@ import (
+	baseDN string
+}
+
+// Error returns the error string for the invalid entry query
```
bad comment

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -10,13 +10,34 @@ import (
 	"github.com/openshift/origin/pkg/cmd/server/api"
 )
 
+// errNoSuchObject is an error that occurs when trying to find a specific entry fails.
```
this looks exactly like `errEntryNotFound`.  In fact, its the same comment.  Why do I have two of these?  defend in the comments or remove.

deads2k commented on pkg/cmd/experimental/syncgroups/cli/prunegroups.go:
```diff
@@ -0,0 +1,234 @@
+func (o *PruneGroupsOptions) Validate() error {
+	results := validation.ValidateLDAPSyncConfig(o.Config)
+	if o.GroupInterface == nil {
+		results.Errors = append(results.Errors, fielderrors.NewFieldRequired("groupInterface"))
```
Inappropriate use of `NewFieldRequired`.  Return an unstructured error saying that its missing.  The `NewFieldRequired` should be reserved for api types are that invalid.

deads2k commented on pkg/cmd/experimental/syncgroups/cli/syncgroups.go:
```diff
@@ -156,82 +157,82 @@ func NewCmdSyncGroups(name, fullName string, f *clientcmd.Factory, out io.Writer
+	return list, nil
+}
+
+func decodeSyncConfigFromFile(configFile string) (*api.LDAPSyncConfig, error) {
```
In usage, you're always dereferencing this.  Make the return consistent with the member or the member consistent with the helper.  I don't care which.

deads2k commented on pkg/cmd/experimental/syncgroups/cli/syncgroups.go:
```diff
@@ -285,6 +286,9 @@ func (o *SyncGroupsOptions) Validate() error {
 
 	results := validation.ValidateLDAPSyncConfig(o.Config)
+	if o.GroupInterface == nil {
+		results.Errors = append(results.Errors, fielderrors.NewFieldRequired("groupInterface"))
```
unstructured error like above

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner_test.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+func checkClientForDeletedGroups(tc *testclient.Fake, expectedGroups []string, t *testing.T) {
+	actualGroups := sets.NewString(extractDeletedGroups(tc)...)
```
convert to set and do the `Difference` in both directions.

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner_test.go:
```diff
@@ -0,0 +1,106 @@
+	"github.com/openshift/origin/pkg/cmd/experimental/syncgroups/interfaces"
+)
+
+func TestGoodPrune(t *testing.T) {
```
Tests like this are a pain, because I have to know the state that various test helpers work with in order to determine whether this test is doing the right thing.  Things that declare their starting and ending data are much easier to follow.

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner_test.go:
```diff
@@ -0,0 +1,106 @@
+	checkClientForDeletedGroups(tc, []string{"os" + Group2UID}, t)
+}
+
+func TestListFailsForPrune(t *testing.T) {
```
There should be tests to be sure existence checks for one group don't fail the whole job

deads2k commented on pkg/cmd/experimental/syncgroups/ad/ldapinterface.go:
```diff
@@ -150,3 +150,20 @@ func (e *ADLDAPInterface) requiredUserAttributes() []string {
 }
+
+// Exists determines if a group idenified with it's LDAP group UID exists on the LDAP server
+func (e *ADLDAPInterface) Exists(ldapGrouUID string) (bool, error) {
```
Add unit tests for all these methods.  Every `Exists` method and be sure to check each branch.

stevekuznetsov replied:
How can we unit-test these? Won't they need an integration test to use the LDAP server mock?

deads2k replied:
> How can we unit-test these? Won't they need an integration test to use the LDAP server mock?

You need to be able to mock response for the queries from `ExtractMembers`.  Looks like either `QueryForEntries` or the `clientConfig.Connect` and `connections.Search`.  You only got to merge without this before because we had to have it for dcut.

stevekuznetsov replied:
Tests added [here](https://github.com/openshift/origin/pull/5145/files#diff-0fc0ddf1eeaf1202d8a291a8604375f7).

deads2k commented:
also, completions are out of date.

stevekuznetsov commented:
@deads2k testing finished
[test][extended:ldap_groups]

stevekuznetsov commented:
Unforseen issue with imports inside of `ldaputil` ... I've refactored the test client placement to fix it. Squashed commits because it's late and making meaningful breaks between commits is hard. 

[test][extended:ldap_groups]

deads2k commented on pkg/auth/ldaputil/ldapclient/interfaces.go:
```diff
@@ -0,0 +1,32 @@
+)
+
+// Client knows how to interact with an LDAP server
+type Client interface {
```
explain the purpose of the interface and describe where you pulled the methods from.

deads2k commented on pkg/auth/ldaputil/ldapclient/interfaces.go:
```diff
@@ -0,0 +1,32 @@
+type Config interface {
+	Connect() (client Client, err error)
+	GetBindCredentials() (bindDN, bindPassword string)
+	DescribeLDAPURL() (schemehostport string)
```
This is icky.  Leave it up the `String()` method.

stevekuznetsov replied:
This is used in `ldaputil` somewhere for an error message... Would be more opaque as `String()` since it's asking for some subset of information from the config.

deads2k replied:
> This is used in ldaputil somewhere for an error message... Would be more opaque as String() since it's asking for some subset of information from the config.

Too icky to keep.

stevekuznetsov replied:
How to refactor this, then? Remove the dependency on the data in the one method that returns the error? Cast the Config in that method?

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -84,17 +85,6 @@ func (a *Authenticator) getIdentity(username, password string) (authapi.UserIden
 		return nil, false, nil
 	}
 
-	// Make the connection and bind to it if a bind DN and password were given
```
These changes shouldn't've been necessary.  Revert them.  You can make a separate pull after the main body of you changes go in.

stevekuznetsov replied:
Almost all of the changes were due to the refactor of `ldapclient.Config` and `ldapclient.Client`. I can revert the two lines that were on top of that to reduce duplicated code if you want, but the rest need to stay.

deads2k commented on pkg/auth/ldaputil/client.go:
```diff
@@ -36,8 +37,8 @@ func NewLDAPClientConfig(URL, bindDN, bindPassword, CA string, insecure bool) (*
-// LDAPClientConfig holds information for connecting to an LDAP server
-type LDAPClientConfig struct {
+// ldapClientConfig holds information for connecting to an LDAP server
+type ldapClientConfig struct {
```
Add the casting equality statement here.

deads2k commented on pkg/auth/ldaputil/client.go:
```diff
@@ -53,14 +54,10 @@ type LDAPClientConfig struct {
 	TLSConfig *tls.Config
 }
 
-func (l LDAPClientConfig) String() string {
```
keep this.

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -225,20 +187,25 @@ func formatResult(results []*ldap.Entry) string {
-	if _, err := clientConfig.Bind(connection); err != nil {
-		return nil, fmt.Errorf("could not bind to the LDAP server: %v", err)
+	if bindDN, bindPassword := clientConfig.GetBindCredentials(); len(bindDN) > 0 {
+		if err := connection.Bind(bindDN, bindPassword); err != nil {
```
@liggitt purity restored.

deads2k commented on pkg/auth/ldaputil/client.go:
```diff
@@ -105,16 +102,10 @@ func (l *LDAPClientConfig) Connect() (*ldap.Conn, error) {
+}
 
-	return false, nil
+func (l *ldapClientConfig) DescribeLDAPURL() string {
```
remove

deads2k commented on pkg/cmd/experimental/syncgroups/locator/grouplocator_test.go:
```diff
@@ -0,0 +1,245 @@
+	return nil, outOfBoundsError
+}
+
+type errEntryNotFoundGenerator struct{}
```
Bad name.  `errEntryNotFound<interfacename>`?

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -131,16 +81,28 @@ func NewLDAPQueryOnAttribute(config api.LDAPQuery, attribute string) (LDAPQueryO
 		if _, err := ldap.ParseDN(attributeValue); err != nil {
 			return nil, fmt.Errorf("could not search by dn, invalid dn value: %v", err)
 		}
+		if !strings.Contains(attributeValue, o.BaseDN) {
```
Why reorder?

stevekuznetsov replied:
If something's happened and we created a malformed DN, the `Contains` check will almost always fail. I think it will be more useful to know that you are about to use a DN that isn't viable first, some of the errors I was able to generate were useless in that they told me my issue was out of bounds, but in reality it was malformed DN.

deads2k replied:
> If something's happened and we created a malformed DN, the Contains check will almost always fail. I think it will be more useful to know that you are about to use a DN that isn't viable first, some of the errors I was able to generate were useless in that they told me my issue was out of bounds, but in reality it was malformed DN.

ok

deads2k commented on pkg/auth/ldaputil/testclient/testclientconfig.go:
```diff
@@ -0,0 +1,37 @@
+	}
+}
+
+func NewConnectingConfig(client ldapclient.Client) ldapclient.Config {
```
This name is weird.  What is `Connecting` about this?  Remove this and just use NewConfig.

stevekuznetsov replied:
I didn't like having empty strings passed in all the time to `NewConfig`, is there a better way to do that?

deads2k replied:
Pass the empty strings.  It makes it easier to follow.

On Fri, Oct 30, 2015 at 8:41 AM, Steve Kuznetsov <notifications@github.com>
wrote:

> In pkg/auth/ldaputil/testclient/testclientconfig.go
> <https://github.com/openshift/origin/pull/5145#discussion_r43497579>:
>
> > +	bindDN       string
> > +	bindPassword string
> > +	ldapURL      string
> > +}
> > +
> > +// NewConfig creates a new Config impl that regurgitates the given data
> > +func NewConfig(client ldapclient.Client, bindDN, bindPassword, ldapURL string) ldapclient.Config {
> > +	return &fakeConfig{
> > +		client:       client,
> > +		bindDN:       bindDN,
> > +		bindPassword: bindPassword,
> > +		ldapURL:      ldapURL,
> > +	}
> > +}
> > +
> > +func NewConnectingConfig(client ldapclient.Client) ldapclient.Config {
>
> I didn't like having empty strings passed in all the time to NewConfig,
> is there a better way to do that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/5145/files#r43497579>.
>

deads2k commented on pkg/auth/ldaputil/testclient/testclientconfig.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+// NewConfig creates a new Config impl that regurgitates the given data
+func NewConfig(client ldapclient.Client, bindDN, bindPassword, ldapURL string) ldapclient.Config {
```
`NewTestConfig`

deads2k replied:
> NewTestConfig

Ok, I see the package.  NewConfig is fine.

stevekuznetsov replied:
It is namespaced. `testclient.NewConfig()` no need to be doubly verbose.

deads2k commented on pkg/auth/ldaputil/testclient/testclient.go:
```diff
@@ -0,0 +1,133 @@
+
+// Fake is a mock client for an LDAP server
+// The following methods define safe defaults for the return values. In order to adapt this test client
+// for a specific test, subtype it and override the method being tested.
```
not subtype.  Anonymously include it and override

deads2k commented on pkg/auth/ldaputil/testclient/testclient.go:
```diff
@@ -0,0 +1,133 @@
+// The following methods define safe defaults for the return values. In order to adapt this test client
+// for a specific test, subtype it and override the method being tested.
+type Fake struct {
+	simpleBindResponse     *ldap.SimpleBindResult
```
is there a good reason for these to private?  private parts of mock objects usually cause me pain.

deads2k commented on pkg/auth/ldaputil/testclient/testclient.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+// NewMatchingSearchErrorClient returns a new matchingSeachError client sitting on top of the parent
+// client. This client returns the gien error when a search base DN matches the given base DN, and
```
type: returns the given error

deads2k commented on pkg/auth/ldaputil/testclient/testclient.go:
```diff
@@ -0,0 +1,133 @@
+
+// matchingSearchErrClient returns the returnErr on every Search() where the search base DN matches the given DN
+// or defers the search to the parent client
+type matchingSearchErrClient struct {
```
Expose these.  I can definitely see tests that want to mutate these values.

deads2k commented on pkg/auth/ldaputil/testclient/testclient.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+// dnMappingClient returns the LDAP entry mapped to by the base dn given, or if no mapping happens, defers to the parent
+type dnMappingClient struct {
```
Expose these. I can definitely see tests that want to mutate these values.

stevekuznetsov replied:
I never ran into any cases ... it's like `PrependReactor`, just create the correct one and stick them together.

deads2k commented on pkg/auth/ldaputil/testclient/testclientconfig.go:
```diff
@@ -0,0 +1,31 @@
+	return tc.client, nil
+}
+
+func (tc *fakeConfig) GetBindCredentials() (string, string) {
```
remove.

stevekuznetsov replied:
This needs to conform to the interface?

deads2k commented on pkg/auth/ldaputil/url.go:
```diff
@@ -246,3 +246,12 @@ func DetermineDerefAliasesBehavior(derefAliasesString string) (DerefAliases, err
 	return derefAliases, nil
 }
+
+func ExtractLDAPHost(url string) (string, error) {
```
I don't like this.  I think I'd rather see an `Host() string` method on the `ldapclient.Config` type.

deads2k commented on pkg/cmd/experimental/syncgroups/ad/augmented_ldapinterface.go:
```diff
@@ -21,7 +22,7 @@ func NewAugmentedADLDAPInterface(clientConfig *ldaputil.LDAPClientConfig,
 		groupQuery:          groupQuery,
 		groupNameAttributes: groupNameAttributes,
-		cachedGroups:        map[string]*ldap.Entry{},
+		cachedGroups:        make(map[string]*ldap.Entry),
```
why?  Its its preference, your preference is ugly.

stevekuznetsov replied:
I did it for consistency with `rfc2307`, I really don't care which is used.

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -84,17 +85,6 @@ func (a *Authenticator) getIdentity(username, password string) (authapi.UserIden
 	}
 
-	// Make the connection and bind to it if a bind DN and password were given
-	l, err := a.options.ClientConfig.Connect()
```
still valid

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -84,17 +85,6 @@ func (a *Authenticator) getIdentity(username, password string) (authapi.UserIden
-	if err != nil {
-		return nil, false, err
-	}
-	defer l.Close()
```
still valid

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -84,17 +85,6 @@ func (a *Authenticator) getIdentity(username, password string) (authapi.UserIden
-	}
-	defer l.Close()
-
-	if _, err := a.options.ClientConfig.Bind(l); err != nil {
```
minor refactor

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -119,27 +109,20 @@ func (a *Authenticator) getIdentity(username, password string) (authapi.UserIden
 	)
 
-	glog.V(4).Infof("searching for %s", filter)
-	results, err := l.Search(searchRequest)
```
still valid

deads2k commented on pkg/cmd/experimental/syncgroups/ad/augmented_ldapinterface.go:
```diff
@@ -77,3 +76,9 @@ func (e *AugmentedADLDAPInterface) requiredGroupAttributes() []string {
+
+// Exists determines if a group idenified with it's LDAP group UID exists on the LDAP server
+func (e *AugmentedADLDAPInterface) Exists(ldapGroupUID string) (bool, error) {
+	groupLocator := locator.NewCompoundLocator(locator.NewGroupBasedLocator(e), locator.NewMemberBasedLocator(e))
```
`locator` is a bad package name.  `grouplocator` is better.

deads2k commented on pkg/cmd/experimental/syncgroups/ad/augmented_ldapinterface.go:
```diff
@@ -40,13 +41,11 @@ type AugmentedADLDAPInterface struct {
 	cachedGroups map[string]*ldap.Entry
 }
 
-var _ ldapinterfaces.LDAPMemberExtractor = &AugmentedADLDAPInterface{}
```
put these back here.

stevekuznetsov replied:
very unclear instructions... was definitely told to move to test files

deads2k commented on pkg/cmd/experimental/syncgroups/ad/augmented_ldapinterface_test.go:
```diff
@@ -0,0 +1,121 @@
+	"github.com/openshift/origin/pkg/cmd/experimental/syncgroups/interfaces"
+)
+
+var _ interfaces.LDAPMemberExtractor = &AugmentedADLDAPInterface{}
```
Congrats on finding a spot where Jordan and I are opposed.  I like these where the types are defined to help people see it. He'll come around eventually.  Put these back in the main source

deads2k commented on pkg/cmd/experimental/syncgroups/ad/ldapinterface_test.go:
```diff
@@ -0,0 +1,206 @@
+	}
+}
+
+var searchErr error = errors.New("generic search error")
```
its like a scavenger hunt.  Make one of these in each test function you're running (doesn't have to be per map entry, per function will do).  It's duplication, but I don't want to be hunting around for this thing.

deads2k commented on pkg/cmd/experimental/syncgroups/ad/ldapinterface_test.go:
```diff
@@ -0,0 +1,206 @@
+
+var searchErr error = errors.New("generic search error")
+
+const (
```
move to the top.

deads2k replied:
Also, I'm not a fan, but I don't care enough to change it.  This is forcing too much structure on the tests because different units are starting to rely on these and they aren't logically required.

deads2k commented on pkg/cmd/experimental/syncgroups/interfaces/interfaces.go:
```diff
@@ -35,3 +35,8 @@ type LDAPGroupListerNameMapper interface {
 }
+
+// LDAPGroupLocator determines if a group identified by an LDAP group UID exists on the LDAP server
+type LDAPGroupLocator interface {
```
`LDAPGroupExistenceChecker` or something.  a `Locator` would tell me where somethign is.

deads2k commented on pkg/cmd/experimental/syncgroups/cli/syncgroups.go:
```diff
@@ -144,7 +145,7 @@ func NewCmdSyncGroups(name, fullName string, f *clientcmd.Factory, out io.Writer
 	cmd.Flags().StringVar(&whitelistFile, "whitelist", whitelistFile, "path to the group whitelist file")
 	cmd.Flags().StringVar(&blacklistFile, "blacklist", whitelistFile, "path to the group blacklist file")
-	// TODO enable this we're able to support string slice elements that have commas
+	// TODO(deads): enable this once we're able to support string slice elements that have commas
```
ha!  optimistic.

deads2k commented on pkg/cmd/experimental/syncgroups/grouppruner.go:
```diff
@@ -0,0 +1,86 @@
+	GroupLocator interfaces.LDAPGroupLocator
+	// Maps an LDAP group enrty to an OpenShift Group's Name
+	GroupNameMapper interfaces.LDAPGroupNameMapper
+	// Allows the Syncer to search for OpenShift Groups
```
bad comment.  Its not a syncer and it doesn't do that.

deads2k commented on pkg/cmd/experimental/syncgroups/locator/grouplocator.go:
```diff
@@ -0,0 +1,70 @@
+	return true, nil
+}
+
+func NewCompoundLocator(locators ...interfaces.LDAPGroupLocator) interfaces.LDAPGroupLocator {
```
Godoc on all these.  Specify this is any exists wins

deads2k commented on pkg/cmd/experimental/syncgroups/locator/grouplocator.go:
```diff
@@ -0,0 +1,70 @@
+		if err != nil {
+			return false, err
+		}
+		exists = exists || opinion
```
Why don't I return on success?

stevekuznetsov commented:
@deads2k Addressed comments

liggitt commented on pkg/auth/ldaputil/client.go:
```diff
@@ -8,10 +8,11 @@ import (
+	"github.com/openshift/origin/pkg/auth/ldaputil/ldapclient"
 )
 
-// NewLDAPClientConfig returns a new LDAPClientConfig
```
I was hoping to avoid changes like this and keep this PR limited to isolated prune changes

deads2k replied:
> I was hoping to avoid changes like this and keep this PR limited to isolated prune changes

This was necessary in order to build unit tests.  Given the tradeoff, I'll take the unit tests.

liggitt commented on pkg/auth/ldaputil/query.go:
```diff
@@ -7,46 +7,29 @@ import (
 	"github.com/openshift/origin/pkg/cmd/server/api"
 )
 
-// errEntryNotFound is an error that occurs when trying to find a specific entry fails.
```
are these changes or straight moves? can we do moves and refactors separately?

deads2k replied:
> are these changes or straight moves? can we do moves and refactors separately from the PR?

Moves, plus minor changes (`New` methods for constructing the errors)

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -91,8 +92,10 @@ func (a *Authenticator) getIdentity(username, password string) (authapi.UserIden
-		return nil, false, err
+	if bindDN, bindPassword := a.options.ClientConfig.GetBindCredentials(); len(bindDN) > 0 {
+		if err := l.Bind(bindDN, bindPassword); err != nil {
+			return nil, false, fmt.Errorf("could not bind to the LDAP server: %v", err)
```
Pass the error through without modification.

deads2k commented on pkg/cmd/experimental/syncgroups/ad/ldapinterface.go:
```diff
@@ -22,18 +24,15 @@ func NewADLDAPInterface(clientConfig *ldaputil.LDAPClientConfig,
 		userNameAttributes:        userNameAttributes,
 		groupMembershipAttributes: groupMembershipAttributes,
-		ldapGroupToLDAPMembers:    map[string][]*ldap.Entry{},
+		cachedGroupedUsers:        map[string][]*ldap.Entry{},
```
This name change is bad.  We're dealing with two different things called groups, you must qualify.  If you're matching code somewhere else, fix it there too.

deads2k commented on pkg/cmd/experimental/syncgroups/ad/ldapinterface.go:
```diff
@@ -94,13 +94,13 @@ func (e *ADLDAPInterface) ListGroups() ([]string, error) {
 // populated, this is a no-op.
 func (e *ADLDAPInterface) populateCache() error {
-	if e.cachePopulated {
+	if len(e.cachedGroupedUsers) > 0 {
```
This check is **not** equivalent.  If you extract members first, then do a list, you've returned the wrong data.  In general I'm with refactors that are equivalent being combined, but refactors that are not equivalent shouldn't be hidden like this.

stevekuznetsov replied:
If you consider what must happen, however, for this bug to manifest, you'll see it should be non-issue unless there's an intern or something building their LDAP syncer incorrectly. In order to `ExtractMembers` you need a UID, and to get UIDs you *may* `ListGroups`. However, if you already have group UIDs (therefore being able to call `ExtractMembers`), you would never `ListGroups`.

deads2k replied:
> If you consider what must happen, however, for this bug to manifest, you'll see it should be non-issue unless there's an intern or something building their LDAP syncer incorrectly. In order to ExtractMembers you need a UID, and to get UIDs you may ListGroups. However, if you already have group UIDs (therefore being able to call ExtractMembers), you would never ListGroups.

Still wrong.  Methods that only work because of the order in which they are called will break someday and it will be frustrating.

deads2k commented on pkg/cmd/experimental/syncgroups/grouplocator/grouplocator.go:
```diff
@@ -0,0 +1,70 @@
+	return true, nil
+}
+
+func NewCompoundLocator(locators ...interfaces.LDAPGroupDetector) interfaces.LDAPGroupDetector {
```
Still needs godoc.  is the intent here: must satisfy all, must satisfy one, fails on first error, ignores errors unless they all fail, ???  There are a lot of options.  I think this is: all are checked, any errors cause a fail, any exists returns true.

deads2k commented on pkg/cmd/experimental/syncgroups/grouplocator/grouplocator.go:
```diff
@@ -0,0 +1,70 @@
+	locators []interfaces.LDAPGroupDetector
+}
+
+func (l *CompoundLocator) Exists(ldapGrouUID string) (bool, error) {
```
typo.  You did it throughout this file.

stevekuznetsov commented:
@deads2k @liggitt broke this up better, can factor into multiple pulls if necessary

stevekuznetsov commented:
test flake on `TestRegistryClientConnectPulpRegistry`.

deads2k commented:
re[test]

deads2k commented:
Saw a refactor that looks unnecessary that I don't remember in the pull before.  This will have to wait until I can review again.  Post-3.1

stevekuznetsov commented:
Closing this in favor of another pull.
